## Performing Arithmetic Operations using Python

Let's begin by using Python as a calculator. You can write and execute Python using a code cell within Jupyter. 

> **Working with cells**: To create a new cell within Jupyter, you can select "Insert > Insert Cell Below" from the menu bar or just press the "+" button on the toolbar. You can also use the keyboard shortcut `Esc+B` to create a new cell. Once a cell is created, click on it to select it. You can then change the cell type to code or markdown (text) using the "Cell > Cell Type" menu option. You can also use the keyboard shortcuts `Esc+Y` and `Esc+M`. Double-click a cell to edit the content within the cell. To apply your changes and run a cell, use the "Cell > Run Cells" menu option or click the "Run" button on the toolbar or just use the keyboard shortcut `Shift+Enter`. You can see a full list of keyboard shortcuts using the "Help > Keyboard Shortcuts" menu option.

Run the code cells below to perform calculations and view their result. Try changing the numbers and run the modified cells again to see updated results. Can you guess what the `//`, `%`, and `**` operators are used for?

In [1]:
2 + 3 + 9

14

In [2]:
99 -73

26

In [3]:
23.54 *-1432

-33709.28

In [4]:
100/7

14.285714285714286

In [7]:
100//7 #modulus

14

In [8]:
100%7

2

In [9]:
5**3

125

As you might expect, operators like `/` and `*` take precedence over other operators like `+` and `-` as per mathematical conventions. You can use parentheses, i.e. `(` and `)`, to specify the order in which operations are performed.

## Solving multi-step problems using variables

Let's try solving the following word problem using Python: 

> A grocery store sells a bag of ice for $1.25 and makes a 20% profit. If it sells 500 bags of ice, how much total profit does it make?

We can list out the information provided and gradually convert the word problem into a mathematical expression that can be evaluated using Python. 

*Cost of ice bag ($)* = 1.25

*Profit margin* = 20% = .2

*Profit per bag ($)* = profit margin * cost of ice bag = .2 * 1.25

*No. of bags* = 500

*Total profit* = no. of bags * profit per bag = 500 * (.2 * 1.25)

In [10]:
500*(0.2*1.25)

125.0

Thus, the grocery store makes a total profit of $125. While this is a reasonable way to solve a problem, it's not entirely clear by looking at the code cell what the numbers represent. We can give names to each of the numbers by creating Python *variables*.

> **Variables**: While working with a programming language such as Python, information is stored in *variables*. You can think of variables as containers for storing data. The data stored within a variable is called its *value*.

In [11]:
cost_of_ice_bag =1.25

In [12]:
profit_margin = 0.2

In [13]:
number_of_bags = 500

In [15]:
profit_per_bag = cost_of_ice_bag*profit_margin
profit_per_bag

0.25

In [16]:
total_profit = number_of_bags*profit_per_bag
total_profit

125.0

In [17]:
print("The grocery store makes a total profit of $", total_profit)

The grocery store makes a total profit of $ 125.0


> **`print`**: The `print` function is used to display information. It takes one or more inputs, which can be text (within quotes, e.g., `"this is some text"`), numbers, variables, mathematical expressions, etc. We'll learn more about variables & functions in the next tutorial.

Creating a code cell for each variable or mathematical operation can get tedious. Fortunately, Jupyter allows you to write multiple lines of code within a single code cell. The result of the last line of code within the cell is displayed as the output. 

Let's rewrite the solution to our word problem within a single cell.

In [18]:
# Store input data in variables
cost_of_ice_bag = 1.25
profit_margin = 0.2
number_of_bags = 500

# Perform the required calculations
profit_per_bag = cost_of_ice_bag*profit_margin
total_profit = number_of_bags*profit_per_bag

# Display the result
print('The grocery store makes a total profit of $', total_profit)

The grocery store makes a total profit of $ 125.0


## Evaluating conditions using Python

Apart from arithmetic operations, Python also provides several operations for comparing numbers & variables.

| Operator    | Description                                                     |
|-------------|-----------------------------------------------------------------|
| `==`        | Check if operands are equal                                     |
| `!=`        | Check if operands are not equal                                 |
| `>`         | Check if left operand is greater than right operand             |
| `<`         | Check if left operand is less than right operand                |
| `>=`        | Check if left operand is greater than or equal to right operand |
| `<=`        | Check if left operand is less than or equal to right operand    |

The result of a comparison operation is either `True` or `False` (note the uppercase `T` and `F`). These are special keywords in Python. Let's try out some experiments with comparison operators.

In [19]:
my_favorite_number = 1
my_least_favorite_number = 5
a_neutral_number = 3

In [20]:
# Equality check - True
my_favorite_number == 1

True

In [21]:
# Equality check - False
my_favorite_number == my_least_favorite_number

False

In [22]:
# Not equal check - True
my_favorite_number != a_neutral_number

True

In [23]:
# Not equal check - False
a_neutral_number != 3

False

In [24]:
# Greater than check - True
my_least_favorite_number > a_neutral_number

True

In [25]:
# Greater than check - False
my_favorite_number > my_least_favorite_number

False

In [26]:
# Less than check - True
my_favorite_number < 10

True

In [27]:
# Less than check - False
my_least_favorite_number < my_favorite_number

False

In [28]:
# Greater than or equal check - True
my_favorite_number >= 1

True

In [29]:
# Greater than or equal check - False
my_favorite_number >= 3

False

In [30]:
# Greater than or equal check - False
my_favorite_number >= 3

False

In [31]:
# Less than or equal check - True
3 + 6 <= 9

True

In [32]:
# Less than or equal check - False
my_favorite_number + a_neutral_number <= 3

False

## Combining conditions with logical operators

The logical operators `and`, `or` and `not` operate upon conditions and `True` & `False` values (also known as *booleans*). `and` and `or` operate on two conditions, whereas `not` operates on a single condition.

The `and` operator returns `True` when both the conditions evaluate to `True`. Otherwise, it returns `False`.

| `a`     | `b`    | `a and b` |
|---------|--------|-----------|
|  `True` | `True` | `True`    |
|  `True` | `False`| `False`   |
|  `False`| `True` | `False`   |
|  `False`| `False`| `False`   |

In [33]:
my_favorite_number

1

In [34]:
my_favorite_number > 0 and my_favorite_number <= 3

True

In [35]:
my_favorite_number < 0 and my_favorite_number <= 3

False

In [37]:
my_favorite_number > 0 and my_favorite_number >= 3

False

In [38]:
True and False

False

In [39]:
True and True

True

The `or` operator returns `True` if at least one of the conditions evaluates to `True`. It returns `False` only if both conditions are `False`.

| `a`     | `b`    | `a or b`  |
|---------|--------|-----------|
|  `True` | `True` | `True`    |
|  `True` | `False`| `True`    |
|  `False`| `True` | `True`    |
|  `False`| `False`| `False`   |


In [42]:
a_neutral_number = 3

In [41]:
a_neutral_number == 3 or my_favorite_number < 0

True

In [43]:
a_neutral_number != 3 or my_favorite_number < 0

False

In [44]:
a_neutral_number != 3 or my_favorite_number < 0

False

In [45]:
my_favorite_number < 0 or True

True

In [46]:
False or False

False

In [47]:
not a_neutral_number == 3

False

In [48]:
not False

True

In [49]:
not True

False

In [50]:
(2 > 3 and 4 <= 5) or not (my_favorite_number < 0 and True)

True

In [51]:
not (True and 0 < 1) or (False and True)

False

In [52]:
not True and 0 < 1 or False and True

False

## Adding text styles using Markdown

Adding explanations using text cells (like this one) is a great way to make your notebook informative for other readers. It is also useful if you need to refer back to it in the future. You can double click on a text cell within Jupyter to edit it. In the edit mode, you'll notice that the text looks slightly different (for instance, the heading has a `##` prefix. This text is written using Markdown, a simple way to add styles to your text. Execute this cell to see the output without the special characters. You can switch back and forth between the source and the output to apply a specific style.

For instance, you can use one or more `#` characters at the start of a line to create headers of different sizes:

# Header 1

## Header 2

### Header 3

#### Header 4

To create a bulleted or numbered list, simply start a line with `*` or `1.`.

A bulleted list:

* Item 1
* Item 2
* Item 3

A numbered list:

1. Apple
2. Banana
3. Pineapple

You can make some text bold using `**`, e.g., **some bold text**, or make it italic using `*`, e.g., *some italic text.* You can also create links, e.g., [a link](https://jovian.ai). Images are easily embedded too:

![](https://i.imgur.com/3gjZMYK.png)

Another really nice feature of Markdown is the ability to include blocks of code. Note that code blocks inside Markdown cells cannot be executed.

```
# Perform the required calculations
profit_per_bag = cost_of_ice_bag * profit_margin
total_profit = number_of_bags * profit_per_bag

# Display the result
print("The grocery store makes a total profit of $", total_profit)

```

You can learn the full syntax of Markdown here: https://learnxinyminutes.com/docs/markdown/